### ISSUE TITLE
tsconfig module: NodeNext has typescript error TS7016 Could not find a declaration file

### ISSUE THREAD

**Issue by perspectivezoom** [Annoraaq/grid-engine#435]
I'm using `GridEngineHeadless` to keep track of character movement serverside in a NodeJS Express Colyseus multiplayer server, and encountered the following type resolution error:

<img width="708" alt="Screen Shot 2023-09-12 at 7 23 57 AM" src="https://github.com/Annoraaq/grid-engine/assets/1837622/13be0281-e2fe-4c63-9419-28dc51a33277">

This error does not appear on my NextJS clientside `GridEngineHeadless` usage (duplicating the server logic for nonlaggy local character movement).

I eventually traced the problem down to the `moduleResolution: "NodeNext"` config in my `tsconfig.json`:
<img width="503" alt="Screen Shot 2023-09-12 at 7 24 21 AM" src="https://github.com/Annoraaq/grid-engine/assets/1837622/e2bba3bf-6cc6-4271-a807-efe7c924fb6b">
If I change the `tsconfig.json` to `"module":"ESNext", "moduleResolution": "Node"` (which is what my frontend uses), `npm run build` works fine.

Much Googling later, I discovered the following Typescript issue: https://github.com/microsoft/TypeScript/issues/49160
> Note that types corresponds to `main` only. `main`/`types` will be completely ignored by Node/TypeScript if `exports` is defined.

I've created a minimal repro at https://github.com/perspectivezoom/tsc-nodenext
<img width="708" alt="Screen Shot 2023-09-12 at 8 02 48 AM" src="https://github.com/Annoraaq/grid-engine/assets/1837622/c20e9606-27b1-40f5-affa-642238718e79">
As before, changing the `tsconfig.json` to `"module":"ESNext", "moduleResolution": "Node"` causes `npm run build` to work.

I think the easiest bugfix here is to have the `package.json` `exports` declaration include a `types` path in the `cjs` `require`, but there are other ways to do it.
```json
  "exports": {
    "require": {
      "types": "./dist/src/GridEngine.d.ts"
      "default": "./dist/GridEngine.min.cjs"
    },
    "import": "./dist/GridEngine.esm.min.js"
  },
```
https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing

Note that this is not a blocking bug; I'm using `declare module "grid-engine";` in my project to bypass the error, but this workaround types `GridEngineHeadless` as `any`, which is not great.

EDIT: I am no longer using `GridEngineHeadless` on the serverside, because I couldn't figure out how to not drop input ticks with bad latency. Switched back to the client telling the server its position instead of inputs, which isn't great from a security perspective but not the worst thing in the world. Man, multiplayer is hard.

**Comment by Annoraaq** [comment]
I tried to reproduce the issue but for me it works as expected. But I probably did not replicate your case 100%.

I used your tsconfig.json (at least the part that you showed above) and compiled a file with the content:

```javascript
import { GridEngineHeadless } from "grid-engine";

console.log(GridEngineHeadless);
```

That worked for me. If I change the "exports" section in Grid Engines package.json as you suggest above it breaks.

If this is still an issue for you, could you share a minimum breaking example on codesandbox or similar?

**Comment by perspectivezoom** [comment]
Here you go. I made no changes other than adding a manual `npm install`, but I don't think I needed to do so.

<img width="1510" alt="Screen Shot 2023-09-17 at 8 45 11 AM" src="https://github.com/Annoraaq/grid-engine/assets/1837622/a7d7d9a7-a440-4de7-9cb0-f1c5172c9632">

https://codesandbox.io/p/sandbox/dazzling-goldwasser-cftd8g
Not sure how long anonymous sandboxes last.
You can reproduce by visiting https://githubbox.com/perspectivezoom/tsc-nodenext (note github => githubbox domain change), going to the Terminal tab, and then converting to a cloud sandbox.

**Comment by Annoraaq** [comment]
Thanks for the example repo! I was finally able to reproduce it. However, I could not yet solve the problem with the suggested fix but I am on it.

**Comment by Annoraaq** [comment]
A small update: I think I got the types working thanks to the link to the TS docs you provided.
Now I am facing the next issue to get it running in new node versions: https://github.com/microsoft/TypeScript/issues/49083

TL/DR: I will have to change all imports in the TS code to use .js file endings, even though there are no such .js files... 😭

**Comment by perspectivezoom** [comment]
> I will have to change all imports in the TS code to use .js file endings, even though there are no such .js files...

That is... pretty disgusting. I would rather you not do that, mostly on moral DevX grounds.

I don't know how comfortable you are with your package bundler, but outputting the `d.ts` files as a sibling seems to be the next best solution, if rejiggering the `exports` didn't work: https://github.com/microsoft/TypeScript/issues/49160#issuecomment-1217296897
> Also, this is not known widely enough: all any package author has to do is put their `.d.ts` files in the same directories as their `.js` files, just like `tsc` emits them by default. The only packages that are dealing with issues are ones that put their types in a separate folder for some reason. `.d.ts` and `.js` files are best friends. If you don’t separate them, you literally don’t have to put anything special at all in the package.json.

As a last resort, the other alternative I thought of was to throw up a `npm install --save-dev @types/grid-engine` package. It's not great, but seems barely less smelly than throwing nonexistent `file.js` import references.

And again, if you don't find a clean solution, feel free to close this out. It's a bit of an odd edge case.

**Comment by Annoraaq** [comment]
> I don't know how comfortable you are with your package bundler, but outputting the d.ts files as a sibling seems to be the next best solution, if rejiggering the exports didn't work: https://github.com/microsoft/TypeScript/issues/49160#issuecomment-1217296897

That is exactly how I finally got it working.

> That is... pretty disgusting. I would rather you not do that, mostly on moral DevX grounds.

It reads like I would eventually have to do it anyway, because the TypeScript team made it pretty clear they will not change it and I think enforcing file endings is part of the ECMA specification...

**Comment by Annoraaq** [comment]
@perspectivezoom, can you confirm that it works now with version 2.38.0?

**Comment by perspectivezoom** [comment]
Yep, I'm getting types now in the backend. It is now correctly complaining that I need to `zod` coerce all of my string inputs into the `Direction` enum now:
<img width="854" alt="Screen Shot 2023-10-15 at 8 26 26 PM" src="https://github.com/Annoraaq/grid-engine/assets/1837622/b490f7f2-2eb3-4d3d-a8d5-1a03540e0f82">

I appreciate the fix @Annoraaq. Closing the issue.

**Comment by Annoraaq** [comment]
Thanks for your help and pointing it out in the first place!

**Comment by perspectivezoom** [comment]
Uh oh. Bumped my nextjs frontend to `2.38.0`, and now it's complaining:
<img width="1467" alt="Screen Shot 2023-10-16 at 7 00 53 AM" src="https://github.com/Annoraaq/grid-engine/assets/1837622/f2e21081-8e47-4afb-98f9-f60777885d50">

Let me know if you need a repo. For now, resolution is to keep frontend at `2.37.0`

**Comment by Annoraaq** [comment]
I could reproduce it.

I just released a patch: v2.38.1

It now works for CJS and ESM for me.

**Comment by perspectivezoom** [comment]
Confirmed that upgrading to `2.38.1` gives me types in both frontend and backend. Closing.
